### ISSUE TITLE
[Performance] Compute shaders implementation discussion

### ISSUE THREAD

**Issue by Supamiu** [Portponky/better-terrain#106]
I'm creating a roguelike game which, like most roguelikes, I'm using procedural generation to create my maps (a combination of cellular automata, Random walk and Perlin Noise), which, in my case, is rendered as map using a set of `TileMapLayer`s and your addon on a dual tilemap setup.

It's really working great and is way more reliable and fastet than default Godot terrains, for sure.

Recently, I have drastically improved performances by using a Compute Shader setup for my Cellular Automata and I went from 5s to 7ms to generate a 256x256 map, which is insane. I then tried to benchmark everything and found that with larger surfaces, BetterTerrain can become quite slow, in my case, we're talking about 7 seconds to compute the structure of this map (not the different ground colors, really just the groune/water parts):
![image](https://github.com/user-attachments/assets/121c8ac5-11cd-4452-84a8-412b48637dce)

7 seconds is still in the ok range for most games and my point here isn't to say it's bad at all, more to see how we could improve it, or at least how I could improve it by using a compute shaders-based approach.

The idea would be to have a way to input peering bits for each cell, so the addon only has to deal with rendering.

Basically, how I see the implementation (Game is addon consumer, BetterTerrain is the addon):

- Game computes Cellular Automata using compute shaders, outputs a texture.
- Game computes peering bits using compute shader, each pixel is then transformed to a color that encodes its peering bits.
- Game reads each pixel to extract peering bits per cell, where 1 pixel is a cell in the map
- Game calls `BetterTerrain.set_peering_bits(cells: Array)`
- BetterTerrain handles rendering based on config, with random cells and other configurations such as decoration

By doing this, peering bits computing would be handled by GPU, which should be pretty much instant compared to feeding cells and then iterating over them to find peering bits.

**Comment by Portponky** [comment]
I have no experience in compute shaders, but it seems like it would work, at least for orthogonal grids.

Better Terrain doesn't map out peering bits per cell. Each cell just holds the terrain type. It holds a large lookup table which scores each tile in a terrain for each cell by checking specific neighbor types against a list of valid types. If that can be done in a compute shader it should be super fast.

**Comment by Portponky** [comment]
It shouldn't be a part of Better Terrain plugin because compute shaders aren't supported on all platforms/rendering devices, and the plugin should remain fully compatible. It could be something similar to https://github.com/Portponky/better-terrain-gde which is a C++ implementation of the same terrain logic. I'd be very interested in seeing a working prototype first, I won't have any time to look into it for the foreseeable future.

**Comment by Supamiu** [comment]
I'll try to make something as I have a good environment for that with a nice testing setup, then I'll showcase it somewhere you can see to discuss how it should be made to have it available for more devs 👍

**Comment by nihiluis** [comment]
Why does this library not use GDExtension by default (https://github.com/Portponky/better-terrain-gde)? Can C# users not use it then?

Mainly curious, no offense intended with the question
